### PR TITLE
Let `@TempDir` fail fast with `File` annotated element and non-default file system temp directory

### LINKED ISSUE #3910: `@TempDir` with `File` annotated elements should fail fast if temp directory file system is not default
`TempDirFactory` implementations based on non-default file systems like [jimfs](https://github.com/google/jimfs) and [memoryfilesystem](https://github.com/marschall/memoryfilesystem) cannot work with `@TempDir` annotated elements of type `File`.

See also the `Path::toFile` [Javadoc](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/nio/file/Path.html#toFile()):

> **Throws:**
>    `UnsupportedOperationException` - if this `Path` is not associated with the default provider

Right now, a test like [the example in the user guide](https://junit.org/junit5/docs/current/user-guide/#writing-tests-built-in-extensions-TempDirectory), but with `tempDir` of type `File`:

```java
@Test
void test(@TempDir(factory = JimfsTempDirFactory.class) File tempDir) {
    // perform test
}

static class JimfsTempDirFactory implements TempDirFactory {

    private final FileSystem fileSystem = Jimfs.newFileSystem(Configuration.unix());

    @Override
    public Path createTempDirectory(AnnotatedElementContext elementContext, ExtensionContext extensionContext) throws IOException {
        return Files.createTempDirectory(fileSystem.getPath("/"), "junit-");
    }

    @Override
    public void close() throws IOException {
        fileSystem.close();
    }

}
```

fails with:

```
org.junit.jupiter.api.extension.ParameterResolutionException: Failed to resolve parameter [java.io.File arg0] in method [void io.github.scordio.jimfs.junit.jupiter.Tests.test(java.io.File)]

	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.UnsupportedOperationException
	at com.google.common.jimfs.JimfsPath.toFile(JimfsPath.java:386)
	... 2 more
```

I would like to improve the user experience, clearly stating that `File` annotated elements are not supported.

As of today, the factory doesn't have easy access to the type of the annotated element.

In my extension for Jimfs, I've implemented a workaround:

https://github.com/scordio/jimfs-junit-jupiter/blob/ac6f085a89368823e7f515f335eb44f21d981447/src/main/java/io/github/scordio/jimfs/junit/jupiter/JimfsTempDirFactory.java#L86-L107

## Deliverables

I see two possible directions:

1. JUnit owns the error handling and asks the factory if `File` annotated elements are supported
    * This could be done with a new default method in `TempDirFactory`, e.g., `boolean supportsFile()`.
3. The error handling happens in the factory
    * `AnnotatedElementContext` should expose the element type somehow so that the factory can build up the error handling around it without having to cast the result of `getAnnotatedElement()`. For example, a new `Class<?> getAnnotatedElementType()` method could be added to `AnnotatedElementContext`.

My preference goes to option 1 to reduce duplication in each `TempDirFactory` implementation.

I'm happy to contribute once a solution is agreed.

---

Relates to:
* scordio/jimfs-junit-jupiter@99d946e3ca99c807fb70e28e48b6c588027f3e0a
* marschall/memoryfilesystem-junit-provider#2

CC @marschall

### PR DESCRIPTION
**Pull request by scordio**
## Overview

Closes #3910.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

### CODE REVIEW

**Comment by scordio** [comment]
@marcphilipp, I don't expect more code changes unless you see something that needs to be updated.

The last point I'm considering is updating the Javadoc of `@TempDir` / `TempDirFactory`. I'll consider it and finalize it in the next few hours.

File: jupiter-tests/src/test/java/org/junit/jupiter/engine/extension/CloseablePathTests.java [hunk 1]
```diff
@@ -95,72 +113,107 @@ void cleanupRoot() throws IOException {
...
 			assertThat(closeablePath.get()).isDirectory();
 
 			delete(closeablePath.get());
 			delete(directory);
 		}
 
+		@DisplayName("succeeds if the factory returns a directory on a non-default file system for a Path annotated element")
 		@Test
+		void factoryReturnsDirectoryOnNonDefaultFileSystemWithPath() throws IOException {
+			TempDirFactory factory = spy(new JimfsFactory());
```
Comment (marcphilipp) [inline jupiter-tests/src/test/java/org/junit/jupiter/engine/extension/CloseablePathTests.java:149]: :question: We're not using the spy in this test, are we?
```suggestion
			TempDirFactory factory = new JimfsFactory();
```

**Review by marcphilipp — APPROVED** [review]

File: jupiter-tests/src/test/java/org/junit/jupiter/engine/extension/CloseablePathTests.java (hunk 1 above)
Comment (scordio) [reply jupiter-tests/src/test/java/org/junit/jupiter/engine/extension/CloseablePathTests.java:149]: Indeed, copy-paste leftover. Thanks for spotting it!

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDir.java [hunk 2]
```diff
@@ -40,11 +40,20 @@
...
+ * <ul>
+ * <li>If the field type or parameter type is neither {@link Path} nor {@link File}.</li>
+ * <li>If a field is declared as {@code final}.</li>
+ * <li>If the temporary directory cannot be created.</li>
+ * <li>If the field type or parameter type is {@code File} and a custom
+ *     {@link TempDir#factory() factory} is used, which creates a temporary directory that does
+ *     not belong to the {@link java.nio.file.FileSystems#getDefault() default file system}.</li>
+ * </ul>
+ *
+ * In addition, a {@code ParameterResolutionException} will be thrown for a
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/io/TempDir.java:56]:
```suggestion
 * {@code @TempDir}.
 * An {@link ExtensionConfigurationException} or a
 * {@link ParameterResolutionException} will be thrown in one of the following
 * cases:
 *
 * <ul>
 * <li>If the field type or parameter type is neither {@link Path} nor
       {@link File}.</li>
 * <li>If a field is declared as {@code final}.</li>
 * <li>If the temporary directory cannot be created.</li>
 * <li>If the field type or parameter type is {@code File} and a custom
 *     {@linkplain TempDir#factory() factory} is used, which creates a temporary
 *     directory that does not belong to the
 *     {@linkplain java.nio.file.FileSystems#getDefault() default file system}.
 * </li>
 * </ul>
 *
 * In addition, a {@code ParameterResolutionException} will be thrown for a
```

**Review by marcphilipp — APPROVED** [review]

**Comment by marcphilipp** [comment]
@scordio Thanks again! 🙂👍

**Pull request merged**